File: intelligence/ai_engine/discovery/schema_inspector.py
"""Schema introspection - discovers all tables, columns, types, constraints, and indexes."""

from typing import Any

import psycopg2
import psycopg2.extras


class SchemaInspector:
    """Discovers all tables, columns, types, constraints, and indexes from PostgreSQL."""

    def __init__(self, database_url: str):
        self.database_url = database_url

    def _connect(self):
        return psycopg2.connect(self.database_url)

    def inspect(self) -> dict[str, Any]:
        return {
            "tables": self._get_tables(),
            "columns": self._get_columns(),
            "foreign_keys": self._get_foreign_keys(),
            "primary_keys": self._get_primary_keys(),
            "unique_constraints": self._get_unique_constraints(),
            "indexes": self._get_indexes(),
            "views": self._get_views(),
            "materialized_views": self._get_materialized_views(),
            "extensions": self._get_extensions(),
            "row_counts": self._get_row_counts(),
        }

    def _get_tables(self) -> list[dict]:
        sql = """
            SELECT table_name, table_type
            FROM information_schema.tables
            WHERE table_schema = 'public'
              AND table_type IN ('BASE TABLE', 'VIEW')
            ORDER BY table_name;
        """
        with self._connect() as conn:
            with conn.cursor(cursor_factory=psycopg2.extras.RealDictCursor) as cur:
                cur.execute(sql)
                return [dict(r) for r in cur.fetchall()]

    def _get_columns(self) -> dict[str, list[dict]]:
        sql = """
            SELECT table_name, column_name, ordinal_position, data_type,
                   udt_name, is_nullable, column_default,
                   character_maximum_length, numeric_precision, numeric_scale
            FROM information_schema.columns
            WHERE table_schema = 'public'
            ORDER BY table_name, ordinal_position;
        """
        result: dict[str, list[dict]] = {}
        with self._connect() as conn:
            with conn.cursor(cursor_factory=psycopg2.extras.RealDictCursor) as cur:
                cur.execute(sql)
                for row in cur.fetchall():
                    table = row["table_name"]
                    if table not in result:
                        result[table] = []
                    result[table].append(dict(row))
        return result

    def _get_foreign_keys(self) -> list[dict]:
        sql = """
            SELECT
                tc.table_name AS source_table,
                kcu.column_name AS source_column,
                ccu.table_name AS target_table,
                ccu.column_name AS target_column,
                tc.constraint_name
            FROM information_schema.table_constraints tc
            JOIN information_schema.key_column_usage kcu
                ON tc.constraint_name = kcu.constraint_name
                AND tc.table_schema = kcu.table_schema
            JOIN information_schema.constraint_column_usage ccu
                ON tc.constraint_name = ccu.constraint_name
                AND tc.table_schema = ccu.table_schema
            WHERE tc.constraint_type = 'FOREIGN KEY'
              AND tc.table_schema = 'public';
        """
        with self._connect() as conn:
            with conn.cursor(cursor_factory=psycopg2.extras.RealDictCursor) as cur:
                cur.execute(sql)
                return [dict(r) for r in cur.fetchall()]

    def _get_primary_keys(self) -> dict[str, list[str]]:
        sql = """
            SELECT tc.table_name, kcu.column_name
            FROM information_schema.table_constraints tc
            JOIN information_schema.key_column_usage kcu
                ON tc.constraint_name = kcu.constraint_name
                AND tc.table_schema = kcu.table_schema
            WHERE tc.constraint_type = 'PRIMARY KEY'
              AND tc.table_schema = 'public';
        """
        result: dict[str, list[str]] = {}
        with self._connect() as conn:
            with conn.cursor(cursor_factory=psycopg2.extras.RealDictCursor) as cur:
                cur.execute(sql)
                for row in cur.fetchall():
                    table = row["table_name"]
                    if table not in result:
                        result[table] = []
                    result[table].append(row["column_name"])
        return result

    def _get_unique_constraints(self) -> dict[str, list[str]]:
        sql = """
            SELECT tc.table_name, kcu.column_name
            FROM information_schema.table_constraints tc
            JOIN information_schema.key_column_usage kcu
                ON tc.constraint_name = kcu.constraint_name
                AND tc.table_schema = kcu.table_schema
            WHERE tc.constraint_type = 'UNIQUE'
              AND tc.table_schema = 'public';
        """
        result: dict[str, list[str]] = {}
        with self._connect() as conn:
            with conn.cursor(cursor_factory=psycopg2.extras.RealDictCursor) as cur:
                cur.execute(sql)
                for row in cur.fetchall():
                    table = row["table_name"]
                    if table not in result:
                        result[table] = []
                    result[table].append(row["column_name"])
        return result

    def _get_indexes(self) -> list[dict]:
        sql = """
            SELECT tablename, indexname, indexdef
            FROM pg_indexes
            WHERE schemaname = 'public'
            ORDER BY tablename, indexname;
        """
        with self._connect() as conn:
            with conn.cursor(cursor_factory=psycopg2.extras.RealDictCursor) as cur:
                cur.execute(sql)
                return [dict(r) for r in cur.fetchall()]

    def _get_views(self) -> list[str]:
        sql = """
            SELECT table_name FROM information_schema.views
            WHERE table_schema = 'public'
            ORDER BY table_name;
        """
        with self._connect() as conn:
            with conn.cursor(cursor_factory=psycopg2.extras.RealDictCursor) as cur:
                cur.execute(sql)
                return [r["table_name"] for r in cur.fetchall()]

    def _get_materialized_views(self) -> list[str]:
        sql = """
            SELECT matviewname FROM pg_matviews
            WHERE schemaname = 'public'
            ORDER BY matviewname;
        """
        with self._connect() as conn:
            with conn.cursor(cursor_factory=psycopg2.extras.RealDictCursor) as cur:
                cur.execute(sql)
                return [r["matviewname"] for r in cur.fetchall()]

    def _get_extensions(self) -> list[dict]:
        sql = "SELECT extname, extversion FROM pg_extension;"
        with self._connect() as conn:
            with conn.cursor(cursor_factory=psycopg2.extras.RealDictCursor) as cur:
                cur.execute(sql)
                return [dict(r) for r in cur.fetchall()]

    def _get_row_counts(self) -> dict[str, int]:
        sql = """
            SELECT relname AS table_name, reltuples::bigint AS approximate_row_count
            FROM pg_class
            WHERE relnamespace = (SELECT oid FROM pg_namespace WHERE nspname = 'public')
              AND relkind = 'r';
        """
        with self._connect() as conn:
            with conn.cursor(cursor_factory=psycopg2.extras.RealDictCursor) as cur:
                cur.execute(sql)
                return {r["table_name"]: r["approximate_row_count"] for r in cur.fetchall()}

File: backend/src/intelligence/discovery/schemaInspector.ts
import { QueryTypes } from 'sequelize';
import { sequelize } from '../../config/database';

export interface TableInfo {
  table_name: string;
  table_type: string;
}

export interface ColumnInfo {
  table_name: string;
  column_name: string;
  data_type: string;
  is_nullable: string;
  column_default: string | null;
  ordinal_position: number;
}

export interface ForeignKey {
  constraint_name: string;
  source_table: string;
  source_column: string;
  target_table: string;
  target_column: string;
}

export interface SchemaSnapshot {
  tables: TableInfo[];
  columns: Record<string, ColumnInfo[]>;
  foreign_keys: ForeignKey[];
  primary_keys: Record<string, string[]>;
  row_counts: Record<string, number>;
}

async function getTables(): Promise<TableInfo[]> {
  const rows = await sequelize.query<TableInfo>(
    `SELECT table_name, table_type
     FROM information_schema.tables
     WHERE table_schema = 'public'
       AND table_type IN ('BASE TABLE', 'VIEW')
     ORDER BY table_name`,
    { type: QueryTypes.SELECT }
  );
  return rows;
}

async function getColumns(tableNames: string[]): Promise<Record<string, ColumnInfo[]>> {
  if (!tableNames.length) return {};
  const rows = await sequelize.query<ColumnInfo>(
    `SELECT table_name, column_name, data_type, is_nullable, column_default, ordinal_position
     FROM information_schema.columns
     WHERE table_schema = 'public'
       AND table_name IN (:tableNames)
     ORDER BY table_name, ordinal_position`,
    { type: QueryTypes.SELECT, replacements: { tableNames } }
  );
  const grouped: Record<string, ColumnInfo[]> = {};
  for (const row of rows) {
    if (!grouped[row.table_name]) grouped[row.table_name] = [];
    grouped[row.table_name].push(row);
  }
  return grouped;
}

async function getForeignKeys(): Promise<ForeignKey[]> {
  const rows = await sequelize.query<ForeignKey>(
    `SELECT
       tc.constraint_name,
       kcu.table_name AS source_table,
       kcu.column_name AS source_column,
       ccu.table_name AS target_table,
       ccu.column_name AS target_column
     FROM information_schema.table_constraints tc
     JOIN information_schema.key_column_usage kcu
       ON tc.constraint_name = kcu.constraint_name
       AND tc.table_schema = kcu.table_schema
     JOIN information_schema.constraint_column_usage ccu
       ON tc.constraint_name = ccu.constraint_name
       AND tc.table_schema = ccu.table_schema
     WHERE tc.constraint_type = 'FOREIGN KEY'
       AND tc.table_schema = 'public'`,
    { type: QueryTypes.SELECT }
  );
  return rows;
}

async function getPrimaryKeys(tableNames: string[]): Promise<Record<string, string[]>> {
  if (!tableNames.length) return {};
  const rows = await sequelize.query<{ table_name: string; column_name: string }>(
    `SELECT kcu.table_name, kcu.column_name
     FROM information_schema.table_constraints tc
     JOIN information_schema.key_column_usage kcu
       ON tc.constraint_name = kcu.constraint_name
       AND tc.table_schema = kcu.table_schema
     WHERE tc.constraint_type = 'PRIMARY KEY'
       AND tc.table_schema = 'public'
       AND kcu.table_name IN (:tableNames)`,
    { type: QueryTypes.SELECT, replacements: { tableNames } }
  );
  const grouped: Record<string, string[]> = {};
  for (const row of rows) {
    if (!grouped[row.table_name]) grouped[row.table_name] = [];
    grouped[row.table_name].push(row.column_name);
  }
  return grouped;
}

async function getRowCounts(tableNames: string[]): Promise<Record<string, number>> {
  if (!tableNames.length) return {};
  // Use pg_class for fast approximate counts
  const rows = await sequelize.query<{ table_name: string; row_count: string }>(
    `SELECT c.relname AS table_name, c.reltuples::bigint AS row_count
     FROM pg_class c
     JOIN pg_namespace n ON n.oid = c.relnamespace
     WHERE n.nspname = 'public'
       AND c.relname IN (:tableNames)
       AND c.relkind = 'r'`,
    { type: QueryTypes.SELECT, replacements: { tableNames } }
  );
  const counts: Record<string, number> = {};
  for (const row of rows) {
    counts[row.table_name] = Number(row.row_count) || 0;
  }
  return counts;
}

export async function inspectSchema(): Promise<SchemaSnapshot> {
  const tables = await getTables();
  const baseTableNames = tables
    .filter((t) => t.table_type === 'BASE TABLE')
    .map((t) => t.table_name);

  const allTableNames = tables.map((t) => t.table_name);

  const [columns, foreign_keys, primary_keys, row_counts] = await Promise.all([
    getColumns(allTableNames),
    getForeignKeys(),
    getPrimaryKeys(baseTableNames),
    getRowCounts(baseTableNames),
  ]);

  return { tables, columns, foreign_keys, primary_keys, row_counts };
}
